File: packages/ts-core/src/time.ts
/**
 * Time utilities
 */

export class TimeUtils {
  /**
   * Get current UTC timestamp in ISO format
   */
  static now(): string {
    return new Date().toISOString();
  }

  /**
   * Parse ISO timestamp to Date
   */
  static parse(timestamp: string): Date {
    return new Date(timestamp);
  }

  /**
   * Check if timestamp is expired
   */
  static isExpired(timestamp: string, maxAgeSeconds: number): boolean {
    const date = new Date(timestamp);
    const ageMs = Date.now() - date.getTime();
    return ageMs > maxAgeSeconds * 1000;
  }

  /**
   * Get seconds elapsed since timestamp
   */
  static secondsSince(timestamp: string): number {
    const date = new Date(timestamp);
    return Math.floor((Date.now() - date.getTime()) / 1000);
  }

  /**
   * Add seconds to timestamp
   */
  static addSeconds(timestamp: string, seconds: number): string {
    const date = new Date(timestamp);
    date.setSeconds(date.getSeconds() + seconds);
    return date.toISOString();
  }

  /**
   * Format duration in human-readable format
   */
  static formatDuration(seconds: number): string {
    if (seconds < 60) return `${seconds}s`;
    if (seconds < 3600) return `${Math.floor(seconds / 60)}m ${seconds % 60}s`;
    
    const hours = Math.floor(seconds / 3600);
    const minutes = Math.floor((seconds % 3600) / 60);
    return `${hours}h ${minutes}m`;
  }
}

File: apps/python/shared/src/ds_shared/time.py
"""
Time utilities for consistent timezone handling
"""

from datetime import datetime, timezone
from typing import Optional


def utc_now() -> datetime:
    """
    Get current time in UTC with timezone info
    
    Returns:
        Current datetime in UTC
    """
    return datetime.now(timezone.utc)


def parse_iso_timestamp(timestamp: str) -> datetime:
    """
    Parse ISO 8601 timestamp string to datetime
    
    Args:
        timestamp: ISO 8601 formatted timestamp
    
    Returns:
        Parsed datetime object
    
    Raises:
        ValueError: If timestamp format is invalid
    """
    return datetime.fromisoformat(timestamp.replace('Z', '+00:00'))


def to_iso_timestamp(dt: datetime) -> str:
    """
    Convert datetime to ISO 8601 string
    
    Args:
        dt: Datetime object
    
    Returns:
        ISO 8601 formatted string
    """
    return dt.isoformat()


def time_ago_seconds(dt: datetime) -> float:
    """
    Calculate seconds elapsed since given datetime
    
    Args:
        dt: Past datetime (must be timezone-aware)
    
    Returns:
        Seconds elapsed
    
    Raises:
        TypeError: If datetime is not timezone-aware
    """
    if dt.tzinfo is None:
        raise TypeError("datetime must be timezone-aware")
    
    return (utc_now() - dt).total_seconds()


def is_expired(dt: datetime, max_age_seconds: float) -> bool:
    """
    Check if datetime is older than max_age_seconds
    
    Args:
        dt: Datetime to check
        max_age_seconds: Maximum age in seconds
    
    Returns:
        True if expired, False otherwise
    """
    return time_ago_seconds(dt) > max_age_seconds
